maintainer/maintainer/ethereum/shared.py

```diff
@@ -94,7 +95,7 @@ async def sign_and_broadcast(
 
     logger.info(f'dispatched transaction {tx_id}')
     if not ignore_result:
-        await _track_tx_result(tx_id)
+        asyncio.ensure_future(_track_tx_result(tx, tx_id))
```

Comment:
I understand why do you want to remove this `await` but I have some doubts. I didn't add it here to guarantee transaction ordering (this is done by the awaits on the upper level) but to make sure we wait until the transaction is actually mined before we trigger subsequent transactions. This way we ensure those subsequent transactions work with a fresh chain state. 

If we remove this `await` I think we may run into the following problem:

1.  We trigger transaction `A` which modifies  some chain state, let's call it `X` 
2. We don't wait for transaction `A` receipt so we trigger transaction `B` whose input consists of chain state `X` Also, suppose transaction `B` calls some code which validates their input on-chain. 
3. Transaction `A` is mined and state `X` is modified
4. Transaction `B` arrives but their input is no longer valid because `X` changed so this transaction is reverted.

IIRC, I observed something similar when the relay invoked `addHeaders` and then `markNewHeaviest` without waiting for the receipt of `addHeaders` transaction. In such case, `markNewHeaviest` reverted because one of their `require` conditions was not satisfied.

Reply:
My reading of the code here is that it doesn't *read* Ethereum state, it only *writes* it. It *reads* Bitcoin state, and writes it to Ethereum. So as long as sequencing of the transactions is preserved, the transactions themselves should be fine. In particular, my reading here was that, before the `await`s, we had no guarantee that `_update_best_digest` would dispatch its transaction with a higher nonce than any preceding `_add_headers` or `_add_diff_change` transactions---it depended on how `asyncio` scheduled those calls. So we could `_update_best_digest` on Ethereum without having up-to-date header/diff-change info, and this would fail.

If that reading is true, then order-of-nonce is all that matters, and waiting for executions in order does not. Does it seem true?

Reply:
Mostly true but the specific part I'm worried about is `_update_best_digest`. Here we prepare the `markNewHeaviest` function's input by calling `contract.get_best_block()` which corresponds to `getBestKnownDigest` chain function and then `find_lca` which calls the `isAncestor` function from `Relay` contract. If you look closer to `isAncestor` you can see it uses `previousBlock` which is in turn modified by `addHeaders` function. There is a high chance this is no longer a problem since we ensure correct nonce order but I just want to draw your attention here and make sure we don't return to the same problems we solved recently.

Reply:
Perfect, that's the piece I missed! I'll give it a deeper look and make sure this behaves as expected, or adjust if not.

Reply:
So I think this would generally be guarded by the [preflight](https://github.com/keep-network/relays/blob/bc881b9b503cbdbfdfd8ff3f398fcbc18132f6d9/maintainer/maintainer/header_forwarder/push.py#L177-L186). The caveat is that if we're moving through older blocks quickly (because we're catching up from our initial parameters, for example), that might not work 100% of the time. In those cases, though, I expect the restart would be sufficient, as it should result in fresher state.

I'm thinking we can make this even safer by changing `'latest'` to `'pending'` [in the `get_best_block` call](https://github.com/keep-network/relays/blob/bc881b9b503cbdbfdfd8ff3f398fcbc18132f6d9/maintainer/maintainer/ethereum/contract.py#L94). Thoughts?

Reply:
Yeah, sounds good!

Reply:
Upon further investigation, in cases where the relay is submitting transactions with insufficient gas price, we end up stacking pending transactions. These transactions have consistently cleared without error eventually, so it doesn't seem we need to change to `pending` in this lookup.